### PR TITLE
Make installing the current firmware version a no-op

### PR DESCRIPTION
**Pull request by DoctorVin**
#### What does this PR do
When Flasher is given a set of firmware, if any individual installed firmware version is equal to what is specified, the task fails with a suggestion to re-run it with `--force`. This is cumbersome in the case of modifying a single firmware in the set and then attempting to re-apply the set to a server, and instead of upgrading a single firmware, the entire set must be force-installed.

This PR changes that behavior be removing any firmware that is already at the desired level from the planned install actions. Thus any firmware already at the desired level will remain so and only the changed firmware will be installed.

### CODE REVIEW

File: internal/worker/task_handler.go [hunk 1]
```diff
@@ -131,8 +121,17 @@ func (h *taskHandler) Run(t sw.StateSwitch, args sw.TransitionArgs) error {
 		return errors.Wrap(ErrSaveTask, ErrTaskTypeAssertion.Error())
 	}
 
+	tctx.Logger.WithFields(logrus.Fields{
+		"condition.id": tctx.Task.ID.String(),
+		"worker.id":    tctx.WorkerID.String(),
+	}).Debug("running the plan")
```
Comment (joelrebel) [inline internal/worker/task_handler.go:127]: style nit: add a newline after the Debug statement

File: internal/worker/task_handler.go [hunk 2]
```diff
@@ -86,27 +90,13 @@ func (h *taskHandler) Plan(t sw.StateSwitch, args sw.TransitionArgs) error {
...
-	// validate task has action plans listed
-	if len(task.ActionsPlanned) == 0 {
-		return false, errors.Wrap(errTaskPlanValidate, "no task actions planned")
-	}
-
-	// validate task context has action statemachines for execution
-	if len(tctx.ActionStateMachines) == 0 {
-		return false, errors.Wrap(errTaskPlanValidate, "task action plan empty")
-	}
+func (h *taskHandler) ValidatePlan(_ sw.StateSwitch, args sw.TransitionArgs) (bool, error) {
```
Comment (joelrebel) [inline internal/worker/task_handler.go:93]: Given this is a noop at this point, should this method be removed completely?

File: internal/worker/task_handler.go [hunk 3]
```diff
@@ -337,3 +359,49 @@ func sortFirmwareByInstallOrder(firmwares []*model.Firmware) {
...
+		currentVersion, ok := invMap[strings.ToLower(fw.Component)]
+		switch {
+		case !ok:
+			hCtx.Logger.WithFields(logrus.Fields{
+				"condition.id": hCtx.Task.ID,
+				"worker.id":    hCtx.WorkerID.String(),
+				"component":    fw.Component,
+				"server.id":    hCtx.Asset.ID.String(),
+			}).Warn("inventory missing component")
+		case currentVersion == fw.Version:
```
Comment (joelrebel) [inline internal/worker/task_handler.go:386]: Suggest using `strings.EqualFold()` here, just in case.

File: Makefile [hunk 4]
```diff
@@ -30,7 +30,7 @@ endif
 		"-X $(LDFLAG_LOCATION).GitCommit=$(GIT_COMMIT) \
          -X $(LDFLAG_LOCATION).GitBranch=$(GIT_BRANCH) \
          -X $(LDFLAG_LOCATION).GitSummary=$(GIT_SUMMARY) \
-         -X $(LDFLAG_LOCATION).Version=$(VERSION) \
+         -X $(LDFLAG_LOCATION).AppVersion=$(VERSION) \
```
Comment (joelrebel) [inline Makefile:33]: Update the ldflag on the `build-linux` command below as well

File: internal/worker/task_handler_test.go [hunk 5]
```diff
@@ -66,3 +71,68 @@ func Test_sortFirmwareByInstallOrde(t *testing.T) {
 
 	assert.Equal(t, expected, have)
 }
+
+func TestRemoveFirmwareAlreadyAtDesiredVersion(t *testing.T) {
```
Comment (joelrebel) [inline internal/worker/task_handler_test.go:75]: tests++ \o/

**Review by joelrebel — COMMENTED** [review]
Thanks for the fix and the included test, left a few minor comments and suggestions.

File: internal/worker/task_handler.go [hunk 6]
```diff
@@ -76,6 +76,10 @@ func (h *taskHandler) Plan(t sw.StateSwitch, args sw.TransitionArgs) error {
 		return errors.Wrap(ErrSaveTask, ErrTaskTypeAssertion.Error())
 	}
 
+	tctx.Logger.WithFields(logrus.Fields{
+		"condition.id": tctx.Task.ID.String(),
```
[the remark this answers is not in the document]
Comment (DoctorVin) [reply internal/worker/task_handler.go]: When we create the entry we specify task id, worker id, asset id and BMC already so we can just drop those fields from here.

File: internal/worker/task_handler.go (hunk 2 above)
Comment (DoctorVin) [reply internal/worker/task_handler.go]: Done.

**Review by joelrebel — APPROVED** [review]
awesome!

**Pull request merged**